tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,517 @@
+) -> Set[Text]:
+  """Gets the output node names of the previous subgraph layer.
+
+  Given a remote op layer, we derive the output node names of the previous
+  subgraph layer. Layers tend to have the following order: subgraph layer,
+  remote op layer, subgraph layer, remote op layer, ...
+
+  Args:
+    remote_op_layer: A set of remote op names for a remote op layer.
+    node_name_to_node_def: A mapping from node names to `NodeDef` protos.
+
+  Returns:
+    A set of output node names of the previous subgraph layer.
+  """
+  previous_subgraph_layer_output_node_names = set()
+
+  for remote_op_name in remote_op_layer:
+    for input_node_name in node_name_to_node_def[remote_op_name].input:
+      input_node = node_name_to_node_def[input_node_name]
+
+      # Assumption: Graph inputs and previous remote op outputs are always
+      #             computed and stored.
+      if _is_placeholder_op(input_node) or _is_remote_op(input_node):
+        continue
+      previous_subgraph_layer_output_node_names.add(input_node_name)
+
+  return previous_subgraph_layer_output_node_names
+
+
+def _get_execution_spec_for_subgraph_layer(
+    graph_def: graph_pb2.GraphDef,
+    graph: tf.Graph,
+    node_name_to_node_def: Mapping[Text, node_def_pb2.NodeDef],
+    previously_visited: Set[Text],
+    output_node_names: Set[Text]
+) -> ExecutionSpec:
+  """Constructs one subgraph layer.
+
+  As discussed in _get_execution_specs(), a subgraph layer contains one or
+  more nodes excluding remote ops. Based on a set of output node names, we
+  traverse toward the ancestors (upward) until encountering a "special" node.
+
+  Here, we traverse upward because each node's node_def contains input names
+  but not output names.
+
+  A "special" node could be either a placeholder node, a remote op, or a node
+  visited by a previous layer. Since it is computed/stored prior to the
+  current subgraph layer, we can treat it as an input of the current subgraph
+  layer.
+
+  Args:
+    graph_def: A `GraphDef` proto for the original graph.
+    graph: A tf.Graph instance for the original graph.
+    node_name_to_node_def: A mapping from node names to `NodeDef` protos.
+    previously_visited: A set of node names from previous subgraph layers.
+    output_node_names: A set of output node names for the current subgraph.
+
+  Returns:
+    An ExecutionSpec representing a subgraph layer.
+  """
+  subgraph = graph_pb2.GraphDef()
+  subgraph.versions.CopyFrom(graph_def.versions)
+  subgraph.library.CopyFrom(graph_def.library)
+
+  queue = collections.deque(output_node_names)
+  visited = set()
+
+  while queue:
+    current_node_name = queue.popleft()
+    current_node = node_name_to_node_def[current_node_name]
+
+    if current_node_name not in visited:
+      visited.add(current_node_name)
+
+      if _is_remote_op(current_node) or _is_placeholder_op(current_node) \
+      or current_node_name in previously_visited:
```

Comment:
nit: use parantheses instead of \\

https://google.github.io/styleguide/pyguide.html#32-line-length

> Do not use backslash line continuation except for with statements [...]

Reply:
Addressed.